keps/0036-kudo-operator-toggle-behavior.md

```diff
@@ -0,0 +1,147 @@
+---
+kep-number: 36
+short-desc: KudoOperator Toggling
+title: Parameter-controlled toggling of KudoOperator tasks
+authors:
+  - "@asekretenko"
+owners:
+  - "@asekretenko"
+editor: @asekretenko
+creation-date: 2021-04-30
+last-updated: 2021-04-30
+status: provisional
+see-also:
+  - KEP-23
+  - KEP-29
+---
+
+# Parameter-controlled toggling of KudoOperator tasks
+
+## Table of Contents
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Interface](#interface)
+     * [Validation](#validation)
+     * [Changing enablingParameter on upgrade](#changing-enablingparameter-on-upgrade)
+  * [Implementation Notes](#implementation-notes)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+<!-- [Tools for generating]: https://github.com/ekalinin/github-markdown-toc -->
+
+## Summary
+
+This KEP aims to add an ability to switch an operator dependency on and off
+via a specified parameter of a parent operator. Namely, as a result of
+implementing this KEP, `KudoOperator` task definition will get an optional field
+containing a name of a boolean parameter (of the parent operator), the value
+of which will determine whether the child operator instance should exist.
+
+## Motivation
+From [#1775](https://github.com/kudobuilder/kudo/issues/1775):
+> Currently, KudoOperator tasks don't support toggle behavior which
+> significantly complicates the handling of the optional dependencies. A good
+> example could be Spark and History Server or Kafka and Schema Registry. The
+> only possible options to handle it today are either include the templates
+> as-is into the parent operator and control the contents at the template level
+> or to implement a job (via Toggle task) which deploys/removes the dependency
+> based on a parameter value. Also, cascading updates/upgrades become
+> complicated in these scenarios.
+
+### Goals
+
+* Add an ability to enforce existence/**non-existence** of the child
+  operator instance (and all its children) via a parent operator parameter,
+  i.e. to toggle the dependency.
+
+* Make parent operators specifying an invalid parameter for toggling fail
+  instance admission (as opposed to simply failing a plan with misconfigured
```

Comment:
This seems off... this isn't a "misconfiguration"... it is how it was designed and built.

Reply:
Or just configured by a user, if we are to allow a behavior currenty existing in a `Toggle` task, where a string parameter that parses as `bool` can be used instead of a boolean parameter.... Nevertheless, we don't need to care about the exact reason here; let me simply remove this.